### PR TITLE
[SYSTEMML-618] Deep Learning DML Library

### PR DESCRIPTION
**Pull request by dusenberrymw**
This introduces a new deep learning DML library, **SystemML-NN**.  SystemML-NN is a layers-based library written in pure DML that contains layers with a simple forward/backward API for affine, spatial convolution, max-pooling, non-linearities (relu, sigmoid, softmax, etc.), dropout, loss functions, other layers, optimizers, and gradient checks.

**SystemML-NN**:
* Layers:
  * Core:
    * Affine
    * Spatial Convolution
    * Max Pooling
  * Nonlinearities:
    * ReLU
    * Sigmoid
    * Softmax
    * Tanh
  * Loss:
    * Cross-entropy loss
    * L1 loss
    * L2 loss
    * Log ("Logistic") loss
  * Regularization:
    * Dropout
    * L1 reg
    * L2 reg
* Optimizers:
  * Adagrad
  * Adam
  * RMSprop
  * SGD
  * SGD w/ Momentum
  * SGD w/ Nesterov Momentum
* Tests:
  * Gradient Checks

The upstream, experimental version of the codebase lives at [https://github.com/dusenberrymw/systemml-nn](https://github.com/dusenberrymw/systemml-nn), and this addition represents the portion of the codebase that is relatively stable.  In particular, vanilla RNNs & LSTMs are currently still in experimental status upstream, but will be added soon.  The idea is to rapidly work on the upstream repo, and then contribute stable pieces to the main SystemML repo.  For now, we will keep this in the `scripts/staging` directory.

### CODE REVIEW

**Comment by deroneriksson** [comment]
:thumbsup::thumbsup: Awesome comments and encapsulation. Very clean code. :smiley:

**Comment by dusenberrymw** [comment]
Thanks, @deroneriksson!

**Comment by dusenberrymw** [comment]
cc @niketanpansare, @mboehm7, @prithvirajsen Interested in any thoughts you may have as well.

**Comment by niketanpansare** [comment]
@dusenberrymw Awesome work. I agree with Deron regarding encapsulation and comments  :+1:

**Comment by niketanpansare** [comment]
@dusenberrymw Any thoughts on the data converters/preprocessing utils for reading existing (image) datasets. For example: Mnist, ImageNet, etc.

**Comment by mboehm7** [comment]
Thanks @dusenberrymw It's probably a good idea to keep this in staging until we have full builtin support for DL. Furthermore, you might want to drop the redundant SystemML-NN path prefix.

**Comment by mboehm7** [comment]
since this library can also serve as an example of how to build dml script libraries, it might be worthwhile to include the vectorized functions for pad_image and im2col too.

**Comment by dusenberrymw** [comment]
@niketanpansare Thanks!  As for the image utils, my preference would be to have those written in Scala/Java & Python as part of `MLContext`.  DML is great for math, but not so great for data wrangling, and especially not with raw bytes ;).  The latter point is important because I don't think we can package MNIST data files directly, and otherwise they have to be processed at the byte level.  For some MNIST examples I'm working on, I just grab them with the TensorFlow utilities.

**Comment by dusenberrymw** [comment]
@mboehm7 Thanks for the feedback.  As for the `SystemML-NN` path prefix, my preference would be to keep it, as it signifies a complete neural net library that just so happens to be distributed in the main SystemML repo.  By keeping it, I'll be able to include documentation and examples directly with the rest of the code.  The goal is to promote others to develop libraries for SystemML.

As for the vectorized functions, that's still on the todo list.  I also want to make sure we improve those parfor loops within functions as well. ;)

**Comment by mboehm7** [comment]
for mnist, you could directly use (1) the infinimnist data generator (http://leon.bottou.org/projects/infimnist), which generates libsvm files, (2) MLlibUtils.loadLibSVMFile to convert to LabeledPoints, and (3) finally our existing converters to convert this to binary block.

**Comment by dusenberrymw** [comment]
@mboehm7 Good idea, and that will work great as part of `MLContext`.

@deroneriksson We should add this as a nice convenience method to `MLContext`.

**Comment by deroneriksson** [comment]
@mboehm7 If tests pass [here](https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/393/), can we merge this today for the 0.10.0 release?

**Comment by mboehm7** [comment]
sure - since we're talking about a library of scripts in staging, this is absolutely fine. Also, there should be no need to run any tests here. Thanks for merging.

**Comment by dusenberrymw** [comment]
Great! I want to check one last thing, and then I'll merge this today when I have a moment.

**Comment by deroneriksson** [comment]
Thank you @mboehm7 and @dusenberrymw !

**Comment by deroneriksson** [comment]
@mboehm7 Should the DML scripts in staging be included in the build artifacts or excluded?

**Comment by deroneriksson** [comment]
(BTW, they will always be included in the -src.tar.gz and -src.zip artifacts.)

**Comment by mboehm7** [comment]
personally, I would not include the staging scripts into the jar as 'staging' refers to experimental status. Having them in the tar is fine though.

**Comment by deroneriksson** [comment]
OK I will  exclude them except from the src.tar.gz and src.zip. Thank you.

**Comment by dusenberrymw** [comment]
Merged!

**Comment by deroneriksson** [comment]
Congrats! This is a great addition to SystemML.

**Comment by dusenberrymw** [comment]
Thanks, @deroneriksson!
